### ISSUE TITLE
How does encoding the circle loops to ints work?

### ISSUE THREAD

**Issue by mrakgr** [agda/cubical#180]
```
{-# OPTIONS --cubical --safe #-}
open import Cubical.Core.Everything

open import Cubical.Foundations.Prelude

open import Cubical.Data.Nat
open import Cubical.Data.Int

data S¹ : Type₀ where
  base : S¹
  loop : base ≡ base

helix : S¹ → Type₀
helix base     = Int
helix (loop i) = sucPathInt i

encode : base ≡ base → Int
encode p = transport (λ i → helix (p i)) (pos zero)

q : encode loop ≡ encode (sym loop ∙ loop ∙ loop)
q = refl
```

I can't wrap my head around this. I can see that if you compose `n` `sucPathInt`s, you would get path that increments `n` times rather than just once when doing the transport. But path constructors themselves can't have such computational context. They are more like AST nodes rather than actual operations.

With AST nodes you could interpret them in order to get the same behavior as for regular operations, but I can't see how this is happening.

```
helix (loop i) = sucPathInt i
```

There is the `sucPathInt` and that answers the question of where it is happening, but I can't see how this connects to everything else.

It is not like transport is iterating over a list and counting the number of `loop` occurrences. In `encode p = transport (λ i → helix (p i)) (pos zero)` won't `p i` always be `base`?

**Comment by Alizter** [comment]
So writing `helix (loop x)` is a bit of a lie. What is actually being done is `ap helix (loop x)`. We know that ap's distribute over concatenated paths, since any other path `base = base` must be built from concatenated paths we see that concatenating loop n times gives us sucPathInt concatenated n times. sucPathInt is the equivalence (turned path via univalence) suc which increments an integer.

You don't need to know anything about the i part of the path. The line
```
helix (loop i) = sucPathInt i
```
translates to something like, applying helix to loop at position i is the same as sucPathInt at position i.

**Comment by Saizan** [comment]
In particular `p i` is not always `base`, it can also be `loop i`, `(loop ∙ loop) i` and so on.

`(loop ∙ loop) i` in particular reduces to an `hcomp ...` and `helix` will compute to something interesting when applied to that.

**Comment by Saizan** [comment]
@Alizter 
> So writing `helix (loop x)` is a bit of a lie. What is actually being done is `ap helix (loop x)`. 

`helix (loop x)`, or `helix (p x)` for any path in the circle `p`, is actually a perfectly fine term in cubical type theory, but yeah in HoTT you would say `ap helix p` instead.

**Comment by Alizter** [comment]
@Saizan I know it is a fine term in cubical type theory. But is it not syntactic sugar for the ap?

**Comment by Saizan** [comment]
@Alizter No, `ap` is not a primitive, you would actually define `ap f p` as `\ i -> f (p i)`.

**Comment by mrakgr** [comment]
@Saizan 
> In particular p i is not always base, it can also be loop i, (loop ∙ loop) i and so on.
> (loop ∙ loop) i in particular reduces to an hcomp ... and helix will compute to something interesting when applied to that.

```
w : ∀ i → (loop ∙ loop) i ≡ base
w i = refl
```
```
E:\TheoremProving\CubicalAgdaStuff\hello-circle.agda:24,7-11
hcomp
(λ { j ((~ i ∨ i) = i1)
       → (λ { (i = i0) → base ; (i = i1) → loop (i1 ∧ j) }) _
   ; j (i1 = i0) → outS (inS (loop i))
   })
(outS (inS (loop i)))
!= base of type S¹
when checking that the expression refl has type
(loop ∙ loop) i ≡ base
```

Yeah, I've noticed that it reduces to `hcomp`. But in this example, at least I can see why the `refl` itself cannot prove the equality.

```
helix : S¹ → Type₀
helix base     = Int
helix (loop i) = sucPathInt i

encode : base ≡ base → Int
encode p = transport (λ i → helix (p i)) (pos zero)
```

In terms of behavior `transport` seems to act like a fold. `pos zero` can only be interpreted as a neutral element, but the rest is more confusing to me.

Imagine if the `encode` was written like this...

```
encode p = transport (λ i el → helix (el i)) (pos zero) p
```

Then I could really interpret this as a fold operation over `p`. I could see `p` as a composite of paths, and `el` as a particular path element and the whole thing would be a lot more understandable to me.

```
encode p = transport (λ i → helix (p i)) (pos zero)
```

But instead what I see here is that the path composite is used directly in the mapping function itself. It feels less than a function and more like a macro.

**Comment by mrakgr** [comment]
Why was it designed like that and not like a regular fold?

**Comment by mrakgr** [comment]
@Alizter 
What you wrote made me realize I was making a certain mental error. I am still a bit confused about it actually. Suppose `p` is `loop`.

```
helix (p i)
helix (loop i)
helix base
Int
```

```
helix (p i)
helix (loop i)
sucPathInt i
```

These are two different potential reduction sequences that could be taken for `loop` and I was operating under the assumption that the first one is valid. But I am getting the sense that for some reason the second one is the one that is actually getting executed.

Why is that?

I mean if `loop` is of type `base ≡ base`, then shouldn't applying `i` to it extract the value in either the left or the right side? Where am I going wrong here?

**Comment by mrakgr** [comment]
...Actually, I see where I am wrong. If `loop` is really `base ≡ base` then why am I able to match on the `i` argument?

**Comment by Alizter** [comment]
OK perhaps I am making the mistake here. `loop i` is really a "point in a path", not a path. So it makes perfect sense to write `f (loop i)`.

**Comment by mrakgr** [comment]
```
e : loop ≡ refl
e p = {!!}
```
```
E:\TheoremProving\CubicalAgdaStuff\hello-circle.agda:26,1-2
loop i != base of type S¹
when checking the definition of e
```

I've had this question for a while now ever since I first saw the lectures by Harper and Bauer weeks ago, but it slipped my mind in the furor.

What is the difference between `loop` and `refl`? It does seem like path constructors have some special properties when they are used directly.

**Comment by Saizan** [comment]
`refl : base ≡ base` is the constant path at `base`, i.e. `\ i -> base`.

`loop` is instead a constructor, for any `r : I` we have `loop r : S¹` which is an element of the circle which is not necessarily equal to `base`. We do have `loop i0 = loop i1 = base` though.

This mimics a "loop" in an actual circle: it starts and ends at a specific point `base` but it's somewhere along the circle otherwise.

**Comment by Alizter** [comment]
@Saizan This seems to be quite a substantial difference between HoTT and cubical type theory. I was under the impression that cubical type theory was a special case of HoTT but now it is not the case.

**Comment by jonsterling** [comment]
If those of us who work on cubical type theories have managed to convey that impression, then we have a lot of work to do in our expository work 😉

**Comment by mrakgr** [comment]
@Alizter What exactly is a substantial difference between HoTT and cubical type theory?

**Comment by Alizter** [comment]
For one, you should be able to write loop = refl in HoTT. I had this "path types and ML identity types are equivalent" idea floating around in my head. If this were the case then loop = refl should be equivalent to something in cubical type theory. But now it turns out that loop i = refl i is not even well typed...

**Comment by Saizan** [comment]
@Alizter I'm not sure what you mean by special case here, you can interpret HoTT into CTT by going along what's in `Cubical.Foundations.HoTT-UF`, which accounts for the computational behaviour of J.

However the point of CTT has always been to have more definitional equalities, to make univalence compute, but maybe you didn't expect CTT to be based on different primitives?

**Comment by Saizan** [comment]
@Alizter both `loop ≡ refl` and `loop i ≡ refl {x = base} i` are well-typed, neither is true though.

**Comment by Alizter** [comment]
@Saizan  I meant loop = refl, not loop i = refl i.

**Comment by Alizter** [comment]
@Saizan  I am trying to understand the error that @mrakgr got when typing that.

**Comment by Saizan** [comment]
@Alizter I'm assuming @mrakgr tried to prove that statement by reflexivity to get that error.

**Comment by Alizter** [comment]
Ah OK all is well then.

**Comment by mrakgr** [comment]
> @Alizter I'm assuming @mrakgr tried to prove that statement by reflexivity to get that error.

Yeah, that is what I tried.

**Comment by mrakgr** [comment]
Before I turn my attention to `helix` again, there is something I am wondering about.

```
{-# OPTIONS --cubical #-}

open import Cubical.Core.Everything
open import Cubical.Foundations.Prelude

open import Cubical.Data.Nat

data ℤ : Set where
  pos : (n : ℕ) → ℤ
  neg : (n : ℕ) → ℤ
  posneg : pos 0 ≡ neg 0

sucℤ : ℤ → ℤ
sucℤ (pos n) = pos (suc n)
sucℤ (neg zero) = pos 1
sucℤ (neg (suc n)) = neg n
sucℤ (posneg i) = pos 1
```

At first I thought this was completely straightforward, but now I see that I misunderstood it. Originally what I thought was that what this was saying is that since `ℤ` has `posneg : pos 0 ≡ neg 0` as a path constructor that would act as an assertion at compile time that `sucℤ (pos 0)` and `sucℤ (neg 0)`.

But if that was really the case, for what reason is the `sucℤ (posneg i) = pos 1` case even necessary? It should be possible to infer whether that is true from the rest of the cases. Also I never really understood where that `i` is coming from in path constructors.

```
sucℤ : ℤ → ℤ
sucℤ (pos zero) = pos 2
sucℤ (pos (suc n)) = pos (suc (suc n))
sucℤ (neg zero) = pos 3
sucℤ (neg (suc n)) = neg n
sucℤ (posneg i) = {!!}
```
```
Goal: ℤ
————————————————————————————————————————————————————————————
i : I
———— Constraints ———————————————————————————————————————————
(pos 2) = (?0 (i = i0)) : ℤ
(pos 3) = (?0 (i = i1)) : ℤ
```

Only now do I see that it is trying to link the two sides.

What is unusual here is that this is what one would get if one was trying to prove an equality. It does not at all resemble destructuring a inductive type. Rather this is where a path type is being created.

Actually there seems to be more at work here...

```
suc 2       suc 3
  |          |
  |          |
  |          |
suc 2 ----- suc 3
```

Assuming you could get this to typecheck, that would not only be proof that the horizontal equalities hold, but also that the vertical ones do as well. Obviously the above cannot be true.

```
suc 1       suc 1
  |          |
  |          |
  |          |
suc 1 ----- suc 1
```

In the original program it is trivial. In the `helix`, it is less so.

```
helix : S¹ → Type₀
helix base     = Int
helix (loop i) = sucPathInt i
```

In this function the equality box would look something like...

```
Int                 Int
 |                   |
 |                   |
 |                   |
sucPathInt i0 ----- sucPathInt i1
```

A few posts ago I wrote that I was confused about something...

```
helix (p i)
helix (loop i)
helix base
Int
```

```
helix (p i)
helix (loop i)
sucPathInt i
```

Namely, that the first one is what I would expect the thing to reduce, but the second one is the thing that seems to be executing. Based on the behavior it seems like the truth should be...

```
helix (p i)
helix (loop i)
helix base
Int
sucPathInt i
```

Because `helix base` needs to equal its path constructor case (for both `i0` and `i1`) hence `Int` can be transported to `sucPathInt i`.

**Comment by mrakgr** [comment]
There are still some mysteries however. I am not sure why `helix` even passes typechecking. Sure, both cases reduce to `Int`, but consider the following.

```
e : Int ≡ Int
e i = {!refl!}
```
```
E:\TheoremProving\CubicalAgdaStuff\hello-circle.agda:27,9-13
_x_66 ≡ _x_66 !=< Set of type Set₁
when checking that the expression refl has type Set
```

It seems that `Int` is not equal to itself in Cubical Agda. Given that this is the case why should `Int` be equal to `sucPathInt i`?

**Comment by mrakgr** [comment]
The other mystery is that the `i` in `λ i → helix (p i)` and the `i` in `helix (loop i) = sucPathInt i` do not necessarily have to do anything with one another.

```
helix (p i)
helix (loop i)
helix base
Int
sucPathInt i
```

Why should the `i` in final step of the reduction sequence be exactly the `i` that we want?

**Comment by mrakgr** [comment]
The last mystery is how functions (such as `helix`) distribute over `hcomp`. 

```
w : ∀ i → (loop ∙ sym loop ∙ loop) i ≡ base
w i = {!refl!}
```

I am not doing this for the sake of proving it, but am just trying to print the body of the left side.

```
E:\TheoremProving\CubicalAgdaStuff\hello-circle.agda:24,9-13
hcomp
(λ { j ((~ i ∨ i) = i1)
       → (λ { (i = i0) → base ; (i = i1) → (sym loop ∙ loop) (i1 ∧ j) }) _
   ; j (i1 = i0) → outS (inS (loop i))
   })
(outS (inS (loop i)))
!= base of type S¹
when checking that the expression refl has type
(loop ∙ sym loop ∙ loop) i ≡ base
```

How would `transport (λ i → helix (p i)) (pos zero)` reduce when `p` is `loop ∙ sym loop ∙ loop`? Knowing that would give me confidence. If I am to use Cubical Agda seriously, I need to understand exactly what the primitives are doing.

**Comment by mrakgr** [comment]
Let me contribute a bit more to this thread. I still have no idea when any of the answers to the questions I posed above are, but now that I've gained a bit experience I decided to go to the torus example again.

```
{-# OPTIONS --cubical #-}

open import Cubical.Core.Everything

open import Cubical.Data.Nat
open import Cubical.Data.Prod

data S¹ : Set where
  base : S¹
  loop : base ≡ base
  
data Torus : Set where
  point : Torus
  line1 : point ≡ point
  line2 : point ≡ point
  square : PathP (λ i → line1 i ≡ line1 i) line2 line2

t2c : Torus → S¹ × S¹
t2c point = (base , base)
t2c (line1 i) = loop i , base
t2c (line2 j) = base , loop j
t2c (square i j) = loop i , loop j

c2t : S¹ × S¹ → Torus
c2t (base , base) = point
c2t (loop i , base) = line1 i
c2t (base , loop j) = line2 j
c2t (loop i , loop j) = square i j
```

The first time I saw this I could not make sense of it, but now at least I can see that the constraints do click. Replacing the body of the `t2c (square i j) = loop i , loop j` with a hole `t2c (square i j) = ?`, I get the following.

```
Goal: S¹ × S¹
————————————————————————————————————————————————————————————
j : I
i : I
———— Constraints ———————————————————————————————————————————
(loop i , base) = (?0 (i = i) (j = i0)) : S¹ × S¹
(loop i , base) = (?0 (i = i) (j = i1)) : S¹ × S¹
(base , loop i) = (?0 (i = i0) (j = i)) : S¹ × S¹
(base , loop i) = (?0 (i = i1) (j = i)) : S¹ × S¹
```

I saw similar things in the past and could not really understand the distinction between `i = i0`, `i = i` and such. I guess it might be a habit from my previous programming experience, but I'd always thought of `i` as being either `i0` or `i1`. Thinking in those terms would lead to confusion in Cubical Agda.

To show that the constraints are satisfied by `loop i , loop j`, let me start by substituting `?0` with it.

```
(loop i , base) = ((loop i , loop j) (i = i) (j = i0)) : S¹ × S¹
(loop i , base) = ((loop i , loop j) (i = i) (j = i1)) : S¹ × S¹
(base , loop i) = ((loop i , loop j) (i = i0) (j = i)) : S¹ × S¹
(base , loop i) = ((loop i , loop j) (i = i1) (j = i)) : S¹ × S¹
```

Now let me apply the substitutions.

```
(loop i , base) = ((loop i , loop j) (j = i0)) : S¹ × S¹
(loop i , base) = ((loop i , loop j) (j = i1)) : S¹ × S¹
(base , loop i) = ((loop i0 , loop j) (j = i)) : S¹ × S¹
(base , loop i) = ((loop i1 , loop j) (j = i)) : S¹ × S¹
```

Another step.

```
(loop i , base) = (loop i , loop i0) : S¹ × S¹
(loop i , base) = (loop i , loop i1) : S¹ × S¹
(base , loop i) = (loop i0 , loop i) : S¹ × S¹
(base , loop i) = (loop i1 , loop i) : S¹ × S¹
```

For the next step, first what I need to keep in mind is that `loop i` simplifies to `loop i` - meaning it stays as it is, but `loop i0 = loop i1 = base`. I verified this yesterday in fact. @Saizan said as much too. This is something that completely slipped my mind while I was just looking at the constraints, but now that I am doing each step by hand, I see that it in fact checks out.

```
(loop i , base) = (loop i , base) : S¹ × S¹
(loop i , base) = (loop i , base) : S¹ × S¹
(base , loop i) = (base , loop i) : S¹ × S¹
(base , loop i) = (base , loop i) : S¹ × S¹
```

Qed.

**Comment by mrakgr** [comment]
Let me do the other example as well. It is a bit more complicated.

```
c2t : S¹ × S¹ → Torus
c2t (base , base) = point
c2t (loop i , base) = line1 i
c2t (base , loop j) = line2 j
c2t (loop i , loop j) = ?
```

```
(line1 i) = (?0 (i = i) (j = i0)) : Torus
(line1 i) = (?0 (i = i) (j = i1)) : Torus
(line2 i) = (?0 (i = i0) (j = i)) : Torus
(line2 i) = (?0 (i = i1) (j = i)) : Torus
```

I'll start by replacing `?0` with `square i j`.

```
(line1 i) = (square i j (i = i) (j = i0)) : Torus
(line1 i) = (square i j (i = i) (j = i1)) : Torus
(line2 i) = (square i j (i = i0) (j = i)) : Torus
(line2 i) = (square i j (i = i1) (j = i)) : Torus
```

This is a bit trickier as there is so uncertainty in how to simplify square. Let me do both steps of substitution at once.

```
(line1 i) = (square i i0) : Torus
(line1 i) = (square i i1) : Torus
(line2 i) = (square i0 i) : Torus
(line2 i) = (square i1 i) : Torus
```

The type of `square` is `PathP (λ i → line1 i ≡ line1 i) line2 line2`. So `square i` is of type `line1 i ≡ line1 i`. Hence `square i i0` and `square i i1` would be `line1 i`. This satisfies the first two constraints. Note that `square i` cannot be reduced to `line2`

Unlike `square i`, `square i0` and `square i1` are both reducible to `line2`. Hence `square i0 i` and `square i1 i` would both be `line2 i`. This would satisfy the last two constraints.

**Comment by Saizan** [comment]
@mrakgr You can get the normal form (warning: can be very big) for a term by using the emacs command `C-c C-n`.

Additionally, the commands `C-c C-,`, `C-c C-t`, and `C-c C-d` can be prefixed by `C-u C-u` to get the normal forms of whatever is displayed.

**Comment by mrakgr** [comment]
It seems `loop i` in the pattern is not there arbitrarily. 

```
c2t : S¹ × ℕ → ℕ
c2t (base , zero) = 1
c2t (base , suc x) = 3
c2t (loop i , zero) = 1
c2t (loop i , suc x) = 3
```

As an experiment, in `c2t` I changed one of the arguments to a `ℕ` along with the result and split on the second one. I wasn't really sure, I thought that maybe `c2t (loop i , zero)` would need to be equal to both `c2t (base , zero)` and `c2t (base , suc x)`. When I realized that the bodies are not linked, I finally realized what the path constructor pattern actually means.

Agda won't let me split on `i` directly, but let me suppose that it is possible for the purpose of illustration. As an example, let me split on `loop i` and substitute it with `loop i0` and `loop i1` by hand.

```
c2t : S¹ × ℕ → ℕ
c2t (base , zero) = 1
c2t (base , suc x) = 3
c2t (loop i0 , zero) = 1
c2t (loop i1 , zero) = 1
c2t (loop i0 , suc x) = 3
c2t (loop i1 , suc x) = 3
```

As was shown previously `loop i0 = loop i1 = base`. So let me use that substitution.

```
c2t : S¹ × ℕ → ℕ
c2t (base , zero) = 1
c2t (base , suc x) = 3
c2t (base , zero) = 1
c2t (base , zero) = 1
c2t (base , suc x) = 3
c2t (base , suc x) = 3
```

Once it is written out like this, it becomes blindingly obvious why the bodies are linked.

It would not make sense to have ...

```
c2t (base , zero) = 1
c2t (base , zero) = 2
```

...but if they are equal then the typechecker can give it a pass.

**Comment by mrakgr** [comment]
@Saizan 

> You can get the normal form (warning: can be very big) for a term by using the emacs command C-c C-n.

For some reason, when I tried to use that I can't get it to work properly. I am not sure why. Maybe I need to put it into the hole directly rather than just give it the expression.

**Comment by mrakgr** [comment]
```
c2t : S¹ × S¹ → Torus
c2t (base , base) = point
c2t (loop i , base) = line1 i
c2t (base , loop j) = line2 j
c2t (loop i , loop j) = square i j
```

Now that I know that the patterns intend, let me go through the rest of the examples in a similar style to what I did before. I did the constraints correct in one of the previous posts, but I hadn't understood where they are coming from at the time.

Split on `i`.

```
c2t (base , loop j) = line2 j
---
c2t (loop i1 , loop j) = square i1 j
c2t (loop i0 , loop j) = square i0 j
```

Simplify the pattern.

```
c2t (base , loop j) = line2 j
---
c2t (base , loop j) = square i1 j
c2t (base , loop j) = square i0 j
```

Simplify the body.

```
c2t (base , loop j) = line2 j
---
c2t (base , loop j) = line2 j
c2t (base , loop j) = line2 j
```

Now the right side...

Split on `j`.

```
c2t (loop i , base) = line1 i
---
c2t (loop i , loop i0) = square i i0
c2t (loop i , loop i1) = square i i1
```

Simplify the pattern.

```
c2t (loop i , base) = line1 i
---
c2t (loop i , base) = square i i0
c2t (loop i , base) = square i i1
```

Simplify the body.

```
c2t (loop i , base) = line1 i
---
c2t (loop i , base) = line1 i
c2t (loop i , base) = line1 i
```

Interesting how the split works concurrently. It is not how usual inductive definitions work. If I had something like `a` , `b` where both were `ℕ`s, I'd first split on `a` to get `zero, b` and `suc a', b`, and then on `b` to get `zero, zero` and `zero, suc b'`, and `suc a', zero` and `suc a', suc b'`.

Here this is more akin to `zero, b` and `suc a', b`, and `a, zero` and `a, suc b'`.

At any rate, let me continue this ....

Split on `j`.

```
c2t (base , base) = point
---
c2t (base , loop i0) = line2 i0
c2t (base , loop i1) = line2 i1
```

Simplify the pattern.

```
c2t (base , base) = point
---
c2t (base , base) = line2 i0
c2t (base , base) = line2 i1
```

Simplify the body.

```
c2t (base , base) = point
---
c2t (base , base) = point
c2t (base , base) = point
```

One last case and then this example is done.

Split on `i`.

```
c2t (base , base) = point
---
c2t (loop i0 , base) = line1 i0
c2t (loop i1 , base) = line1 i1
```

Simplify the pattern.

```
c2t (base , base) = point
---
c2t (base , base) = line1 i0
c2t (base , base) = line1 i1
```

Simplify the body.

```
c2t (base , base) = point
---
c2t (base , base) = point
c2t (base , base) = point
```

Qed.

**Comment by mrakgr** [comment]
Let me do the other example as well.

```
t2c : Torus → S¹ × S¹
t2c point = (base , base)
t2c (line1 i) = loop i , base
t2c (line2 j) = base , loop j
t2c (square i j) = loop i , loop j
```

Split on `i`.

```
t2c (line2 j) = base , loop j
---
t2c (square i0 j) = loop i0 , loop j
t2c (square i1 j) = loop i1 , loop j
```

Simplify the patterns.

```
t2c (line2 j) = base , loop j
---
t2c (line2 j) = loop i0 , loop j
t2c (line2 j) = loop i1 , loop j
```

Simplify the bodies.

```
t2c (line2 j) = base , loop j
---
t2c (line2 j) = base , loop j
t2c (line2 j) = base , loop j
```

Now the other side.

Split on `j`.

```
t2c (line1 i) = loop i , base
---
t2c (square i j0) = loop i , loop j0
t2c (square i j1) = loop i , loop j1
```

Simplify the patterns.

```
t2c (line1 i) = loop i , base
---
t2c (line1 i) = loop i , loop j0
t2c (line1 i) = loop i , loop j1
```

Simplify the bodies.

```
t2c (line1 i) = loop i , base
---
t2c (line1 i) = loop i , base
t2c (line1 i) = loop i , base
```

Now come the other two cases.

Split on `i`.

```
t2c point = (base , base)
---
t2c (line1 i0) = loop i0 , base
t2c (line1 i1) = loop i1 , base
```

Simplify the patterns.

```
t2c point = (base , base)
---
t2c point = loop i0 , base
t2c point = loop i1 , base
```

Simplify the bodies.

```
t2c point = (base , base)
---
t2c point = base , base
t2c point = base , base
```

Now comes the last case.

Split on `j`.

```
t2c point = (base , base)
---
t2c (line2 i0) = base , loop i0
t2c (line2 i1) = base , loop i1
```

Simplify the patterns.

```
t2c point = (base , base)
---
t2c point = base , loop i0
t2c point = base , loop i1
```

Simplify the bodies.

```
t2c point = (base , base)
---
t2c point = base , base
t2c point = base , base
```

Qed.

**Comment by mrakgr** [comment]
Let me just check one of the isomorphism sides.

```
c2t-t2c : (t : Torus) → c2t (t2c t) ≡ t
c2t-t2c point = refl
c2t-t2c (line1 i) = refl
c2t-t2c (line2 j) = refl
c2t-t2c (square i j) = refl
```

Let me replace `refl` with `?`. 

```
c2t-t2c : (t : Torus) → c2t (t2c t) ≡ t
c2t-t2c point = ?
c2t-t2c (line1 i) = ?
c2t-t2c (line2 j) = ?
c2t-t2c (square i j) = ?
```

I'll go through each of the holes in turn. Actually, I am going to have to do some unsimplification by hand. The upper half are the relevant cases (rather than the goal) and the lower half are the reduction steps. Let me go from each of the holes in turn.

Here is the first one.

```
t2c point = (base , base)
c2t (base , base) = point

---

c2t (t2c point) ≡ point
c2t (base , base) ≡ point
point ≡ point
```

Here is the second one.

```
t2c (line1 i) = loop i , base
c2t (loop i , base) = line1 i

---

c2t (t2c (line1 i)) ≡ line1 i
c2t (loop i , base) ≡ line1 i
line1 i ≡ line1 i
```

Here is the third one.

```
t2c (line2 j) = base , loop j
c2t (base , loop j) = line2 j

---

c2t (t2c (line2 j)) ≡ line2 j
c2t (base , loop j) ≡ line2 j
line2 j ≡ line2 j
```

Here is the last one.

```
t2c (square i j) = loop i , loop j
c2t (loop i , loop j) = square i j

---

c2t (t2c (square i j)) ≡ square i j
c2t (loop i , loop j) ≡ square i j
square i j ≡ square i j
```

Qed.

**Comment by mrakgr** [comment]
With this let me go back to the `encode` example. It is a bit more clear to me now.

```
helix : S¹ → Type₀
helix base     = Int
helix (loop i) = sucPathInt i

encode : base ≡ base → Int
encode p = transport (λ i → helix (p i)) (pos zero)
```

I wasn't really sure how `helix (loop i)` would reduce. I knew that the `helix (loop i) = sucPathInt i` was being executed, but the path constructor pattern did not really make sense. I thought that `loop i` would have have to reduce to `base`, but path constructors have some weird rules that aren't obvious at first glance.

I was not sure how to interpret their pattern and I kept wondering where the `i` in `helix (loop i) = ...` is coming from. When you construct a path using `loop` you never actually pass it `i`. So it did not make sense to me to be able to destructure it. Moreover you can't actually take apart regular paths as they are functions. Path constructors act weird in that they seem to be both paths and inductive types.

Now I understand enough to know that...

```
helix (p i)
helix (loop i)
helix base
Int
sucPathInt i
```

...is definitely not a valid reduction sequence. The `loop i` is definitely not being reduced to `base` and then converted to `sucPathInt i` for some strange reason.

```
helix (p i)
helix (loop i)
sucPathInt i
```

Instead this is what is going on and this answers a few of the mysteries. Some still remain though.

I still do not understand how transport distributes over homogeneous composition. Or is it `helix` itself that is getting distributed before being passed to `transp`? I have no idea.

How does reduction for this work?

**Comment by mrakgr** [comment]
> We will note, however, that for HITs hcomp (λ i → λ { (r = i1) → u}) u0 only reduces to u[i1 / i] when r = i1, and is to be considered a canonical element otherwise. Therefore, functions defined by pattern matching on a HIT also have to make progress when provided an element built with hcomp. We will often refer to such an element as hcomp r u u0 for ease of notation. The next section describes how this can be achieved, in the context of a core type theory for Cubical Agda.

What does this mean?

> Our main technical contribution is an elaboration algorithm for (co)pattern matching definitions in the presence of HITs and path applications. Following Cockx and Abel [2018] we formulate our algorithm as a translation from (co)pattern matching clauses to case trees. The main challenges are generating the computational behavior on hcomp elements of HITs and making sure clauses for path constructors agree with what the function does at the endpoints of the path.

> Recalling that hcomp r u u0 is also a canonical element of the circle, we can see that we additionally have to cover the following cases: ...

So `hcomp` is a canonical element. It seems that on page 18 there are of how it reduces. This is just what I wanted.

Though I was reading the documentation before, I skipped this as the parts that came previously lost me so I was not even aware that the section existed.

**Comment by mrakgr** [comment]
The section on page 18 of the Cubical Agda paper only half a page, but what is there is enough to satisfy me. I now have a basic, but usable model of how the composite paths reduce. With this I understand HITs well enough to use them and I am done with what would be the tutorial stage for Cubical Agda.

There were a lot of steps that I had to take to get to this point and hopefully the [issues I've opened](https://github.com/agda/cubical/issues?utf8=%E2%9C%93&q=author%3Amrakgr+) on this repo should be of assistance to other beginners. It definitely took some effort. Ultimately, a lot of what I've discovered should be in the documentation and stated explicitly rather than having newcomers play Sherlock Holmes.

If I find Cubical Agda useful for some of the things that I want to do, I will do a tutorial to help improve the state of it. Either way that will be far into the future though and for now the plan is to continue studying the Cubical library. Hopefully I should have an easier time now.

Thank you for bearing with me.
